syncapi/sync/requestpool.go

```diff
@@ -37,18 +40,63 @@ import (
 // RequestPool manages HTTP long-poll connections for /sync
 type RequestPool struct {
 	db       storage.Database
+	cfg      *config.SyncAPI
 	userAPI  userapi.UserInternalAPI
 	notifier *Notifier
 	keyAPI   keyapi.KeyInternalAPI
 	rsAPI    roomserverAPI.RoomserverInternalAPI
+	lastseen sync.Map
 }
 
 // NewRequestPool makes a new RequestPool
 func NewRequestPool(
-	db storage.Database, n *Notifier, userAPI userapi.UserInternalAPI, keyAPI keyapi.KeyInternalAPI,
+	db storage.Database, cfg *config.SyncAPI, n *Notifier,
+	userAPI userapi.UserInternalAPI, keyAPI keyapi.KeyInternalAPI,
 	rsAPI roomserverAPI.RoomserverInternalAPI,
 ) *RequestPool {
-	return &RequestPool{db, userAPI, n, keyAPI, rsAPI}
+	rp := &RequestPool{db, cfg, userAPI, n, keyAPI, rsAPI, sync.Map{}}
+	go rp.cleanLastSeen()
+	return rp
+}
+
+func (rp *RequestPool) cleanLastSeen() {
+	for {
+		rp.lastseen.Range(func(key interface{}, value interface{}) bool {
+			lastseen := value.(time.Time)
+			if time.Since(lastseen) > time.Minute*2 {
+				rp.lastseen.Delete(key)
+			}
+			return true
+		})
+		time.Sleep(time.Minute)
+	}
+}
+
+func (rp *RequestPool) updateLastSeen(req *http.Request, device *userapi.Device) {
+	value, _ := rp.lastseen.LoadOrStore(device.UserID+device.ID, time.Now().Add(-time.Minute))
+	lastseen := value.(time.Time)
+	if time.Since(lastseen) < time.Minute {
+		return
+	}
+
+	remoteAddr := req.RemoteAddr
```

Comment:
Doesn't this mean if you hit the server with the same device but different IPs we drop all but 1? Is that an issue?

Reply:
It will, but I don't know what the alternative is. If we check if the IP address changed OR the time is over a minute, then two or more devices syncing at the same time will just result in the value flapping and hitting the user API a lot more often.